servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/GrpcRequestTracker.java

```diff
@@ -0,0 +1,244 @@
+import io.servicetalk.transport.api.TransportObserver;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.loadbalancer.RequestTracker.REQUEST_TRACKER_KEY;
+
+final class GrpcRequestTracker {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GrpcRequestTracker.class);
+    private static final GrpcLifecycleObserver.GrpcRequestObserver NOOP_REQUEST_OBSERVER =
+            new NoopGrpcRequestObserver();
+
+    private static final Function<GrpcStatus, ErrorClass> PEER_RESPONSE_ERROR_CLASSIFIER = (status) -> {
+        // TODO: this needs to be gone over with more detail.
+        switch (status.code()) {
+            case OK:
+                return null;
+            case CANCELLED:
+                return ErrorClass.CANCELLED;
+            default:
+                return ErrorClass.EXT_ORIGIN_REQUEST_FAILED;
+        }
+    };
+
+    // TODO: this needs to be gone over with more detail.
+    private static final Function<Throwable, ErrorClass> ERROR_CLASS_FUNCTION = (exn) ->
+            ErrorClass.EXT_ORIGIN_REQUEST_FAILED;
+
+    private GrpcRequestTracker() {
+        // no instances
+    }
+
+    static <ResolvedAddress> ConnectionFactoryFilter<ResolvedAddress, FilterableStreamingHttpConnection> filter() {
+        return (connection) -> new ConnectionFactoryWrapper<>(connection);
+    }
+
+    private static class ConnectionFactoryWrapper<ResolvedAddress>
+            extends DelegatingConnectionFactory<ResolvedAddress, FilterableStreamingHttpConnection> {
+
+        ConnectionFactoryWrapper(
+                ConnectionFactory<ResolvedAddress, FilterableStreamingHttpConnection> delegate) {
+            super(delegate);
+        }
+
+        @Override
+        public Single<FilterableStreamingHttpConnection> newConnection(
+                ResolvedAddress resolvedAddress, @Nullable ContextMap context, @Nullable TransportObserver observer) {
+            Single<FilterableStreamingHttpConnection> result = delegate()
+                    .newConnection(resolvedAddress, context, observer);
+            if (context == null) {
+                LOGGER.debug("Context is null. In order for {} to get access to the {}" +
+                                ", health-monitor of this connection, the context must not be null.",
+                        GrpcRequestTracker.class.getSimpleName(), RequestTracker.class.getSimpleName());
+            } else {
+                result = result.map(connection -> transformConnection(connection, context));
+            }
+            return result;
+        }
+
+        private FilterableStreamingHttpConnection transformConnection(
+                FilterableStreamingHttpConnection connection, ContextMap context) {
+            RequestTracker requestTracker = context.remove(REQUEST_TRACKER_KEY);
```

Comment:
Will it make sense to remove `RequestTracker` from the `context` even before calling `delegate().newConnection(...)`?  That way we can skip `map` completely for users who don't have default LB enabled.

Reply:
I did try that at first but switched to this so we can use filter ordering to pick the 'winning' tracker. The actual connection factory wrapping happens on the return path so we want our filter to be later in seeing the connection so the tracker is closer to the LB. However, seeing the connection late means that it sees the context early so if we wanted to get the `RequestTracker` before delegating then we'd lose that ability to override since the HttpRequestTracker would always get first dibs.